### PR TITLE
Define basic interfaces for using Merkle tree

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: common/src/lib.rs [hunk 1]
```diff
@@ -99,3 +99,24 @@ impl BlockHeader {
...
+    // TODO
+}
+
+#[derive(Error, Debug, Serialize, Deserialize, Clone)]
+pub enum MerkleProofError {
+    /// When the proof is malformed.
+    #[error("malformed proof: {0}")]
+    MalformedProof(String),
+    /// When the root doesn't match
+    #[error("unmatched string: found {0} but expected {1}")]
```
Comment (TomTaehoonKim) [inline common/src/lib.rs:114]: How about `expected {1}, but found {0}`? I think I've seen more of the error messages like this.

File: merkle-tree/src/lib.rs [hunk 2]
```diff
@@ -1 +1,60 @@
...
+pub enum Error {
+    /// When the given storage is not a valid Merkle-tree-indexed database.
+    #[error("invalid storage {0}")]
+    MalformedStorage(String),
+    /// When the provided storage emitted an error during the operation.
+    #[error("storage error: {0}")]
+    StorageError(simperby_kv_storage::Error),
+}
+
+/// A state machine that interpret the given storage as a Merkle tree rooted by a specific key.
```
Comment (TomTaehoonKim) [inline merkle-tree/src/lib.rs:18]: interpret -> interprets

File: merkle-tree/src/lib.rs [hunk 3]
```diff
@@ -1 +1,60 @@
...
+///
+/// Note that there is no `get()` method as leaf nodes (which the user of this struct would like to read)
+/// are stored by their keys directly, so one can just access directly to the underlying `KVStore` using the key.
```
Comment (TomTaehoonKim) [inline merkle-tree/src/lib.rs:21]: KVStore -> KVStorage

File: merkle-tree/src/lib.rs [hunk 4]
```diff
@@ -1 +1,60 @@
...
+pub struct MerkleTree<S> {
+    // TODO: add the fields (for real) and remove _x (dummy)
+    _x: PhantomData<S>,
+}
+
+impl<S: KVStorage> MerkleTree<S> {
```
Comment (posgnu) [inline merkle-tree/src/lib.rs:27]: Maybe we need a sort of `contain(value) -> bool` method?

File: merkle-tree/src/lib.rs [hunk 5]
```diff
@@ -1 +1,60 @@
...
+    #[error("invalid storage {0}")]
+    MalformedStorage(String),
+    /// When the provided storage emitted an error during the operation.
+    #[error("storage error: {0}")]
+    StorageError(simperby_kv_storage::Error),
+}
+
+/// A state machine that interpret the given storage as a Merkle tree rooted by a specific key.
+///
+/// Note that there is no `get()` method as leaf nodes (which the user of this struct would like to read)
```
Comment (posgnu) [inline merkle-tree/src/lib.rs:20]: I don't understand this part. Isn't it that the `MerkleTree` is an upper-level abstraction upon `KVstorage`? In that case, I believe `get()` method is still needed unless our PBC code holds both `KVstorage` and `MerkleTree`.

File: kv-storage/src/lib.rs [hunk 6]
```diff
@@ -1,27 +1,36 @@
...
+    #[error("unknown: {0}")]
+    Unknown(String),
+}
 
 /// TODO: Provide error types.
 ///
 /// Note: This trait is quite subject to change.
 #[async_trait]
-pub trait KVStore {
+pub trait KVStorage {
```
Comment (posgnu) [inline kv-storage/src/lib.rs:17]: We need `contain()` here?

File: simperby/src/lib.rs [hunk 7]
```diff
@@ -16,7 +16,7 @@ pub trait BlockExecutor {
     type Transaction;
     async fn execute(
         &self,
-        store: &mut dyn KVStore,
+        store: &mut dyn KVStorage,
```
Comment (Milchstra3e) [inline simperby/src/lib.rs:19]: it's better to change from "store" to "storage"

**Review by Milchstra3e — DISMISSED** [review]
checked

File: merkle-tree/src/lib.rs (hunk 4 above)
Comment (junha1) [reply merkle-tree/src/lib.rs:27]: fixed.

File: kv-storage/src/lib.rs (hunk 6 above)
Comment (junha1) [reply kv-storage/src/lib.rs:17]: no, for the same reason with `get()`

File: merkle-tree/src/lib.rs (hunk 4 above)
Comment (junha1) [reply merkle-tree/src/lib.rs:27]: no, for the same reason with `get()`.

File: kv-storage/src/lib.rs (hunk 6 above)
Comment (junha1) [reply kv-storage/src/lib.rs:17]: fixed.

File: merkle-tree/src/lib.rs (hunk 3 above)
Comment (junha1) [reply merkle-tree/src/lib.rs]: fixed.

File: merkle-tree/src/lib.rs (hunk 2 above)
Comment (junha1) [reply merkle-tree/src/lib.rs]: fixed.

File: merkle-tree/src/lib.rs (hunk 5 above)
Comment (junha1) [reply merkle-tree/src/lib.rs]: No it's not. MerkleTree is just one way of interpreting the given storage. We still need to access to the underlying storage for the following cases.
1. We need to access to entries that are not part of the Merkle tree. (e.g., past blocks)
2. We need to use the 'checkpoint' functionalities provided by the storage.
That makes it natural not to provide `get()` method that would just invoke another `get()` of the underlying storage.

Comment (junha1) [reply merkle-tree/src/lib.rs]: I add a commit about this. Please check https://github.com/postech-dao/simperby/pull/39/commits/a6d9a653000896de8a3fd56efbb5eaf78c222c54

File: kv-storage/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    /// Reverts all the changes made since the last checkpoint.
+    async fn revert_to_latest_checkpoint(&mut self) -> Result<(), Error>;
+    /// Inserts a key-value pair into the storage. If exists, it will be overwritten.
+    async fn insert_or_update(&mut self, key: Hash256, value: &[u8]) -> Result<(), Error>;
+    /// Removes a key-value pair from the storage. If not exists, it will fail.
+    async fn remove(&mut self, key: Hash256) -> Result<(), Error>;
+    /// Retrieves the value associated with the key. If not exists, it will return `None`.
+    async fn get(&self, key: Hash256) -> Result<Option<Vec<u8>>, Error>;
+    /// Checks whether the given item exists in the storage.
+    async fn contain(&self, key: Hash256) -> Result<(), Error>;
```
Comment (Milchstra3e) [inline kv-storage/src/lib.rs:37]: how contain fn notify its result?
I thk, contain(val) -> bool is better.

Comment (junha1) [reply kv-storage/src/lib.rs:37]: @Milchstra3e Same as `get`
i.e., `Ok(())` if exists, `Error` if 
1. not exists
2. DB error
Error variants will be added later.

File: merkle-tree/src/lib.rs [hunk 9]
```diff
@@ -1 +1,70 @@
...
+    pub async fn get_size_overhead(&self) -> Result<usize, Error> {
+        unimplemented!()
+    }
+
+    /// Creates a Merkle proof for the given entry.
+    pub async fn create_merkle_proof(&self, _key: Hash256) -> Result<MerkleProof, Error> {
+        unimplemented!()
+    }
+
+    /// Returns a reference to the underlting storage.
```
Comment (lemonholic) [inline merkle-tree/src/lib.rs:61]: typo

File: kv-storage/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    where
+        Self: Sized;
+    /// Open an existing storage with the path given.
+    async fn open(path: &str) -> Result<Self, Error>
+    where
+        Self: Sized;
+    /// Records the current state to the persistent storage.
+    async fn commit_checkpoint(&mut self) -> Result<(), Error>;
+    /// Reverts all the changes made since the last checkpoint.
+    async fn revert_to_latest_checkpoint(&mut self) -> Result<(), Error>;
```
Comment (lemonholic) [inline kv-storage/src/lib.rs:29]: Can we commit checkpoint two times in a row and then revert to latest two times?
In other words, will the commits stack up if we call `commit_checkpoint` multiple times?

**Review by posgnu — DISMISSED** [review]

File: merkle-tree/src/lib.rs (hunk 9 above)
Comment (junha1) [reply merkle-tree/src/lib.rs]: fixed

File: kv-storage/src/lib.rs (hunk 10 above)
Comment (junha1) [reply kv-storage/src/lib.rs:35]: stated.

File: kv-storage/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    /// Inserts a key-value pair into the storage. If exists, it will be overwritten.
+    async fn insert_or_update(&mut self, key: Hash256, value: &[u8]) -> Result<(), Error>;
+    /// Removes a key-value pair from the storage. If not exists, it will fail.
+    async fn remove(&mut self, key: Hash256) -> Result<(), Error>;
+    /// Retrieves the value associated with the key. If not exists, it will return `None`.
+    async fn get(&self, key: Hash256) -> Result<Option<Vec<u8>>, Error>;
```
[the remark this answers is not in the document]
Comment (junha1) [reply kv-storage/src/lib.rs]: fixed.

**Review by lemonholic — APPROVED** [review]

**Pull request merged**